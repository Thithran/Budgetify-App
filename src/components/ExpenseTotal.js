import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import numeral from 'numeral';

export class ExpenseTotal extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            sum: 0,
        }
    }
    componentDidMount = () => {
        let sum = 0;
        this.props.expenses.map((expense) => {
            sum = sum + expense.amount;
        })
        this.setState({ sum });
    }
    render() {
        return (
            <div>
                Total of expenses is: {numeral(this.state.sum / 100).format('$0,0.00')}
            </div>
        )
    }
}

export default ExpenseTotal;
