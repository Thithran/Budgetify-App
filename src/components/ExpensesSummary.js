import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpenseSummary = (props) => {
    const expenseWord = props.expensesCount === 1 ? 'expense' : 'expenses';
    const formattedExpensesTotal = 'â‚¬' + numeral(props.expensesTotal / 100).format('0,0.00')
    console.log(props.expensesTotal)
    return (
        <div>
            <h2>
            {!(props.expensesCount === 0) && <div>Viewing {props.expensesCount} {expenseWord} with total: {formattedExpensesTotal}</div>}
            </h2>
        </div>
    )
};
const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expensesCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    };
  };
  
  export default connect(mapStateToProps)(ExpenseSummary);
  