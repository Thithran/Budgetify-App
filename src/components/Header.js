import React from 'react';
import { NavLink} from 'react-router-dom';

const Header = () => (
    <header>
        <h1>Budgetify</h1>
        <NavLink exact to="/" activeClassName="active">Home page </NavLink>
        <NavLink to="/create" activeClassName="active">Create </NavLink>
    </header>
)

export default Header;