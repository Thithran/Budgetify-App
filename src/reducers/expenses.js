// When an command is called, Actions handels all the information, then sent it to Reducers to perform it 

// command -> Translated by Actions -> Perform by reducers 

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE': 
            return [
                ...state,
                action.expense
            ]; // this return a new list of expenses

        case 'REMOVE_EXPENSE': 
            return state.filter(({id}) => id !== action.id);

        case 'EDIT_EXPENSE': 
            return state.map((expense)=> {
                if (expense.id === action.id){
                    return {
                        ...expense, 
                        ...action.updates
                    }
                } else {
                    return expense;
                }
            })
        default:
            return state;
    }
};

