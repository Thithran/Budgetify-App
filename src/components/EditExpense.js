import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense, removeExpense} from '../actions/expenses'
const EditExpensePage = (props) => {
    return (
        <div>
            <ExpenseForm 
                expense={props.expense}
                
                onSubmit={(expense) => {
                    // Dispatch the action to edit the expense
                    // Redirect to the dashboard
                    props.dispatch(editExpense(props.expense.id, {...expense}));
                    props.history.push('/');
                    // console.log('updated', expense);
                }}
            />
            <button onClick={() => {
                    props.dispatch(removeExpense({id: props.expense.id}))
                    props.history.push('/');
                    // console.log('removed', expense)
                }}>Remove</button>
        </div>
    )
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    }
};
export default connect(mapStateToProps)(EditExpensePage);
// store.dispatch (editExpense(expenseTwo.expense.id, {amount: 500}));

// Technique: Create onSubmit tools and onClick button 
// Reuse ExpenseForm to edit 