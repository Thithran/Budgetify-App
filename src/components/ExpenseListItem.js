// Export a stateless functional component 
// description, amount, createdAt 

import React from 'react';
import {Link} from 'react-router-dom'; // you have to use LINK in order to remain the ID for expense. a href automatically refresh new ID 
import {connect} from 'react-redux';
import { removeExpense } from '../actions/expenses'
import moment from 'moment';
import numeral from '../style/numeral';

export const ExpenseListItem = ({id, dispatch, description, amount, createdAt}) => (
    <div>
        <Link to= {`/edit/${id}`}>
        <h3>{description}</h3></Link> 
        <p>
        {numeral(amount)}  {` `}
        - 
        {` `}{moment(createdAt).format('MMMM Do, YYYY')}
        </p>
        <button onClick={() => {
            dispatch(removeExpense({id}))
        }} >Remove</button>
    
    </div>
);

export default connect()(ExpenseListItem);


