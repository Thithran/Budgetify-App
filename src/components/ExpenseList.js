import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import expenseSelector from '../selectors/expenses'
export const ExpenseList = (props) => (
    <div>
    {   
        props.expenses.length === 0 ? (
            <p>No expenses</p>
        ) : (
            props.expenses.map(el => 
            <ExpenseListItem 
                key={el.id}
                {...el}
            />)
        )
    }
    {}
    </div>
);

const ConnectedExpenseList = connect((state) => {
    return {
        expenses: expenseSelector        (state.expenses, {...state.filters})    
    };
    
})(ExpenseList);

export default ConnectedExpenseList;
