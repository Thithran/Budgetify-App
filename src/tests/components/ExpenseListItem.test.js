import React from 'react';
import expenses from '../fixtures/expenses';
import { shallow } from 'enzyme';
import { ExpenseListItem } from '../../components/ExpenseListItem';

test ('Should render Item with fixture data', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[1]}/>);
    expect(wrapper).toMatchSnapshot();
}) 