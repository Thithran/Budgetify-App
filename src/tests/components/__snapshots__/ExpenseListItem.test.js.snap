// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Item with fixture data 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/2"
  >
    <h3>
      rent
    </h3>
  </Link>
  <p>
    â‚¬0,01
      
     
    -
     
    December 28th, 1969
  </p>
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
